crypto/x509/x_pubkey.c

```diff
@@ -263,9 +263,43 @@ int i2d_PUBKEY(const EVP_PKEY *a, unsigned char **pp)
     if (a->ameth != NULL && a->ameth->pub_encode != NULL
         && !a->ameth->pub_encode(xpk, a))
         goto error;
-    xpk->pkey = (EVP_PKEY *)a;
-    ret = i2d_X509_PUBKEY(xpk, pp);
-    xpk->pkey = NULL;
+    if (a->ameth != NULL) {
+        xpk->pkey = (EVP_PKEY *)a;
+        ret = i2d_X509_PUBKEY(xpk, pp);
+        xpk->pkey = NULL;
+        goto error;
+    }
+    {
+        const char *serprop = OSSL_SERIALIZER_PUBKEY_TO_DER_PQ;
+        OSSL_SERIALIZER_CTX *ctx =
+            OSSL_SERIALIZER_CTX_new_by_EVP_PKEY(a, serprop);
+        BIO *out = BIO_new(BIO_s_mem());
+
+        if (ctx != NULL
+            && out != NULL
+            && OSSL_SERIALIZER_CTX_get_serializer(ctx) != NULL
+            && OSSL_SERIALIZER_to_bio(ctx, out)) {
+            BUF_MEM *buf = NULL;
+
+            BIO_get_mem_ptr(out, &buf);
+            ret = buf->length;
+
+            if (pp != NULL) {
+                if (*pp == NULL) {
```

Comment:
This code looks very strange..
In the NULL case *pp points to the start of the buffer
In the other case it increments the buffer by ret.
Not sure how this works unless the calling code is doing equivalent code for the NULL case??

Reply:
From the manual:

> **i2d_TYPE**() encodes the structure pointed to by _\*a_ into DER format. If _ppout_ is not NULL, it writes the DER encoded data to the buffer at _\*ppout_, and increments it to point after the data just written. If the return value is negative an error occurred, otherwise it returns the length of the encoded data.
> 
> If _\*ppout_ is NULL memory will be allocated for a buffer and the encoded data written to it. In this case _\*ppout_ is not incremented and it points to the start of the data just written.

Reply:
You really should read the manual.  `perldoc -o man doc/man3/d2i_X509.pod`.  It has NOTES and WARNINGS sections that say quite a bit

Reply:
yep - I know why its done (Do I think its good - not really :))
Thanks for the info : d2i_X509.pod is not the first place I would think of to look.